### ISSUE TITLE
ng-content projection for root component

### ISSUE THREAD

**Issue by RomanGotsiy** [angular/angular#4946]
`ng-content` projection doesn't work for root component. E.g. for the following code:

```
<app>
   <p> Some content that needs to be projected </p>
</app>

@Component({
  selector: 'app',
})
@View({
  template: `<p>Hello world</p>
              <ng-content></ng-content>`
})
class App {}

bootstrap(App);
```

only `<p> Hello world <p>` persists in the DOM after `App` is bootstrapped.

I've reviewed the [design doc](https://docs.google.com/document/d/1BXUFVpImyWN7ynyPKehAesctP0UJTaDYbDTNVkarSeQ/edit) and haven't found such restriction.

[Plunker](http://plnkr.co/edit/10aPDcz0EUML1rjK7S3V?p=preview)

**Comment by vicb** [comment]
see https://github.com/angular/angular/issues/1858

**Comment by mhevery** [comment]
This should be an easy fix. We just have to pass in the content into the `componentFactory.create` [here](https://github.com/angular/angular/blob/master/modules/@angular/core/src/application_ref.ts#L456).

/cc @tbosch Do you think this is reasonable?

**Comment by vicb** [comment]
note: we do not compile the index (only cmp templates)

**Comment by mhevery** [comment]
@vicb true, I don't think this feature is about compiling, it is only about projecting the nodes, which should be easy.

**Comment by DzmitryShylovich** [comment]
@mhevery 
> We just have to pass in the content into the componentFactory.create here.

The question is how to access that content? Something like this would work
```
const rootNode = document.querySelector(componentFactory.selector);
    if (!rootNode) {
      throw new Error(`The selector "${componentFactory.selector}" did not match any elements`);
    }
    const childNodes = Array.prototype.slice.call(rootNode.childNodes);
    const compRef = componentFactory.create(this._injector, [childNodes], rootNode);
```
but we don't have access to the DOM at this point (Renderer is not available yet).

**Comment by tbosch** [comment]
@mhevery What about `<ng-content select="...">`? We don't have the css selector any more during runtime, not even the attribute `select`...

So we would need to somehow restrict this feature to only work for `<ng-content>` elements that have no `select` attribute, and throw for others. This requires somehow marking these `ng-content` elements during compilation.

**Comment by tbosch** [comment]
Also a component can reproject nodes, so we really need to have the list of nodes.

What could work is the following:

- store the NgContent selectors in the view definition
- add a new method `queryDirectChildren` to the renderer
- make `bootstrap` call this method, get the matching children and then call the regular `ComponentFactory.create`.

However, this is a bigger change and I am not sure we should do this now, as it would also make `ApplicationRef.bootstrap` asynchronous (as the method on the `Renderer` needs to return a Promise for WebWorkers).

**Comment by mhevery** [comment]
OK, let's do it post v4.

**Comment by colthreepv** [comment]
Is there any known workaround to use this functionality until this gets a proper fix?

**Comment by tbosch** [comment]
Note that this is a bit easier now, as we introduced `ComponentFactory.ngContentSelectors` (in master, not yet in an rc (see [here](https://github.com/angular/angular/blob/master/packages/core/src/linker/component_factory.ts#L76)). This is e.g. used by ng-upgrade [here](https://github.com/angular/angular/blob/master/packages/upgrade/src/common/downgrade_component_adapter.ts#L184) to fill the `projectableNodes` for the `ComponentFactory.create` call.

**Comment by leon** [comment]
Any news on this?

**Comment by kirkgrover** [comment]
Once support for <ng-content> is added for bootstrapped/root components, will any components in the projected content be rendered?  For example: 

```
<app>
    <h1>App Heading</h1>
    <component-one></component-one>
</app>
```

In this example, if app's template is '<ng-content></ng-content>', and component-one's template is '<div>App Content</div>', would the rendered HTML be:

```
<app>
    <h1>App Heading</h1>
    <div>App Content</div>
</app>
```

Sorry if this was already answered in what @vicb and @mhevery said about compiling.  I'm trying to make the jump from AngularJS to Angular, and I'm seeking clarity on the direction of this feature as I plan how or if we'll be able to use Angular in our next project.

Many thanks!

**Comment by megamindbrian** [comment]
Any update on this?

**Comment by MCKRUZ** [comment]
Angular team, can you provide an update on when/if this is going to be resolved.  I have a large client that we are having to provide a pretty clunky workaround for in order to circumvent this.  Is it on the roadmap?

**Comment by msp166** [comment]
Has any progress been made on this issue?

**Comment by Rush** [comment]
So is it really working fine with Ivy?

**Comment by Hobart2967** [comment]
> So is it really working fine with Ivy?

no. unfortunately not.

**Comment by Hobart2967** [comment]
> Any update on this?
> 
> What is the recommended style to accomplish this?
> 
> ```
> <span *ngIf="!routerLink"><ng-content></ng-content></span>
> <a *ngIf="routerLink" md-button [routerLink]="routerLink" (click)="click.emit()">
>     <ng-content></ng-content>
> </a>
> ```
> 
> Currently `<ng-content>` just shows up blank probably because it is being inserted in to the wrong place or not at all.

dunno if you still need this, but utilize a ng-template to reduce the number of ng-content usages to exactly one, then reuse the template which doesnt count into the projectionNodes index :

```
<ng-template #content>
   <ng-content>
   </ng-content>
</ng-template>
<a [href]="uri | async" *ngIf="isExternal | async" [target]="target | async">
  <ng-template [ngTemplateOutlet]="content"></ng-template>
</a>
<a [routerLink]="uri | async" *ngIf="!(isExternal | async)" [queryParams]="queryParams | async" [target]="target | async">
  <ng-template [ngTemplateOutlet]="content"></ng-template>
</a>
```

**Comment by yahyaKacem** [comment]
Any update on this?
is it possible now to project content into root element?

**Comment by RichFinn-WTW** [comment]
> I don't think this is going to be fixed. Since the functionality is available through angular elements instead. Angular elements also solves having @inputs on root components, which is another thing you want when integrating into a cms

We use angular elements in our cms integration specifically for this reason of projections and @inputs. Works great

**Comment by jklappstein** [comment]
My company has come to the end of the road of a long conversion from angularjs to angular and now, at the end, we are a bit stuck. We are serving pages as razor views in .net core MVC and in those views we arbitrarily used angular tags. Now, we have no way of passing through the RenderBody() content to the app-root (whereas before it didn't matter where the tags were in relation to app-root). Content projection would have made out life so easy but until this is fixed we seem very stuck. Maybe someone can give us some guidance in terms of any possible work around short of completely rewriting the UI.

**Comment by lakmal-gentem** [comment]
> My company has come to the end of the road of a long conversion from angularjs to angular and now, at the end, we are a bit stuck. We are serving pages as razor views in .net core MVC and in those views we arbitrarily used angular tags. Now, we have no way of passing through the RenderBody() content to the app-root (whereas before it didn't matter where the tags were in relation to app-root). Content projection would have made our life so easy but until this is fixed we seem very stuck. Maybe someone can give us some guidance in terms of any possible work around short of completely rewriting the UI.

On app init, you can parse the DOM for those content wrapped elements and store them in a service and access them from components...

**Comment by RichFinn-WTW** [comment]
@jklappstein - as I mentioned in my previous comment, we use Angular Elements to get around the limitation of the root element not supporting content projection for this exact reason. In an older asp.net webforms app, where we also output 'random' angular component tags, we used to just bootstrap the DOM in AnguarJS and everything worked. Now we use Angular Elements to do something like this:

_.aspx (webform):
```html
<custom-ngelement-tag>
  <span class="projection_slot_1">
      <div id="serversideOutput" runat="server"></div>
      <!-- everything that might be angularish is output into this server-side tag -->
  </span>
</custom-ngelement-tag>
```
_.component.html (ng component tied to element):
```html
<ng-content select=".projection_slot_1"></ng-content><!-- normal content projection -->
```

translating the webform snippet above to MVC, your _Layout.cshtml could look like this:
```html
<custom-ngelement-tag>
  <span class="projection_slot_1">
      @RenderBody()
  </span>
</custom-ngelement-tag>
```

Lastly, Elements are pretty easy to create:
```ts
const ngCustomElement= createCustomElement(MyComponent, { injector });//MyComponent could be your root component or some other component in your bundle - doesn't matter.
customElements.define('custom-ngelement-tag', ngCustomElement);
```
